### PR TITLE
Support for  `upload_multiple_artifacts`

### PR DESCRIPTION
**Pull request by eds-collabora**
This provides support for the gitlab runner feature tag `upload_multiple_artifacts`, and more generally brings artifact handling into closer alignment with the Gitlab runner. There are several parts to this:

- Client crates specify the files they can provide, rather than taking responsibility for constructing an output zip.
- Gzip format as well as ZIP format artifacts are supported.
- User selection of desired artifacts is supported, including some limited support for globbing via the `glob` crate.

A major driver here is that if a client crate is able to provide junit test results, supporting this feature tag and exposing that file now provides integration with the gitlab pipelines test tab feature, and without the client crate having to be aware of this in any deep way.

### CODE REVIEW

File: gitlab-runner/src/client.rs [hunk 1]
```diff
@@ -22,9 +22,58 @@ where
...
 const JOB_STATUS: &str = "Job-Status";
 
-#[derive(Debug, Clone, Serialize)]
+fn is_false(value: &bool) -> bool {
+    !value
+}
+
+#[derive(Debug, Default, Clone, Serialize)]
 struct FeaturesInfo {
+    #[serde(skip_serializing_if = "is_false")]
```
Comment (sjoerdsimons) [inline gitlab-runner/src/client.rs:31]: rather then is_false you can use `std::ops::Not::not` directly

File: gitlab-runner/src/run.rs [hunk 2]
```diff
@@ -54,36 +55,74 @@ where
...
+            continue;
+        }
+        if let Ok(mut uploader) = Uploader::new(client.clone(), &build_dir, response.clone()) {
+            let mut available_files: Vec<U> = handler.get_uploadable_files().await?;
+            let mut res = Ok(());
+            let mut uploaded = 0;
+
+            for path in artifact.paths.iter() {
+                let matches = match glob::Pattern::new(path) {
+                    Ok(pattern) => available_files
```
Comment (sjoerdsimons) [inline gitlab-runner/src/run.rs:84]: This is may cause files to be matched and uploaded multiple times.. This should really be turned around so for each artifacts it filters if any of the paths match and then uploads

File: gitlab-runner/src/lib.rs [hunk 3]
```diff
@@ -106,6 +106,36 @@ macro_rules! outputln {
...
+/// generated or downloaded from some other source on demand. The
+/// `get_path` method is required so that the globbing Gitlab expects
+/// can be performed without the handler needing to be involved.
+pub trait UploadableFile: Eq {
+    /// Get the logical path of the file.
+    ///
+    /// This is the path on disk from the root of the checkout for
+    /// Gitlab's own runner. It should match the paths that users are
+    /// expected to specify when requesting artifacts in their jobs.
+    fn get_path(&self) -> String;
```
Comment (sjoerdsimons) [inline gitlab-runner/src/lib.rs:122]: why `String` instead of `&str`

**Review by sjoerdsimons — CHANGES_REQUESTED** [review]

File: gitlab-runner/src/run.rs (hunk 2 above)
Comment (eds-collabora) [reply gitlab-runner/src/run.rs:84]: @sjoerdsimons I'm not sure I follow here. The `available_files` mess (obviously not elegant) takes care of multiple uploads (we delete things from `available_files` when we upload them for a given artifact). But the fact that the same file can be uploaded multiple times _for_different_artifacts_ is required, this is an important part of matching Gitlab's behaviour. For example, if you specify `junit.xml` in reports and under the normal artifacts, it is uploaded twice. The first time it's going into the zip file (that's one artifact), and the second time it's getting gzipped and bypassing the normal pathways inside Gitlab, so that it enters the web interface.

The actual "upload_multiple_artifacts" feature tag means we can repackage data multiple times in different ways to make Gitlab happy. It describes to us what it wants in each one, but they are isolated from each other.

File: gitlab-runner/src/run.rs [hunk 4]
```diff
@@ -54,36 +55,74 @@ where
...
+                        .iter()
+                        .enumerate()
+                        .filter(|(_, f)| pattern.matches(&f.get_path()))
+                        .map(|(i, _)| i)
+                        .rev()
```
[the remark this answers is not in the document]
Comment (eds-collabora) [reply gitlab-runner/src/run.rs:89]: It's a little evil, this really would be much nicer with `drain_filter` if it was stabilized. But this allows us to iterate over the indices, removing them from `available_files`, and not invalidate the other indices: since they're in reverse order, processing tail first keeps all other indices correct until we remove them.

File: gitlab-runner/src/run.rs [hunk 5]
```diff
@@ -54,36 +55,74 @@ where
...
-        if let Ok(mut uploader) = Uploader::new(client, &build_dir, response) {
-            let r = handler.upload_artifacts(&mut uploader).await;
-            if r.is_ok() {
-                uploader.upload().await.and(script_result)
-            } else {
-                r
+    let mut overall_result = script_result;
+
+    for artifact in response.artifacts.iter().take(1) {
+        let upload = !cancel_token.is_cancelled()
```
[the remark this answers is not in the document]
Comment (eds-collabora) [reply gitlab-runner/src/run.rs:61]: Will fix. Obviously the different artifacts have different upload behaviour, but cancellation should not play a part here, my mistake.

File: gitlab-runner/src/lib.rs (hunk 3 above)
Comment (eds-collabora) [reply gitlab-runner/src/lib.rs:122]: I did it this way because for the runner I was working with, it's convenient to generate the path names on demand (as e.g. `format!("{}_log.yaml", self.job_id)`) and making it a `&str` gives us headaches there (it effectively forces us to store the string somewhere, and there's no particular reason to). 

In general, I assume for non-standard runners, it's quite likely that you aren't really reading files off the filesystem, you're handing out promises to download or generate something from somewhere else, if the path you register it under is requested by the job. In that case, it's convenient to generate the filename on demand.

However, if there's an efficiency concern about repeatedly generating the filenames here, I could easily force the client crate to store them?

Comment (sjoerdsimons) [reply gitlab-runner/src/lib.rs:122]: Yeah the user could keep a cache of UploadFiles in that case indeed with an allocated string. Though likely it's not going to be a real concern

File: gitlab-runner/src/run.rs (hunk 2 above)
Comment (sjoerdsimons) [reply gitlab-runner/src/run.rs:84]: Oh i see you use a mutable vector and then remove the ones you uploaded.. ugh that's hard to parse as a code flow

File: gitlab-runner/src/run.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sjoerdsimons) [reply gitlab-runner/src/run.rs:89]: uhggg; yeah please try to avoid that; would be better to get an iterator over matching files and upload those rather then fool around with indexes whichis way too subtle

File: gitlab-runner/src/lib.rs (hunk 3 above)
Comment (sjoerdsimons) [reply gitlab-runner/src/lib.rs:122]: If you want to easily support both cases you could make it Cow<str>  as well ofcourse

**Review by sjoerdsimons — CHANGES_REQUESTED** [review]
Seems to be shaping up nicely ; would be good to have a test-case for multiple artifact upload as well (and in multiple formats) so we can unit test this code

File: gitlab-runner/src/lib.rs [hunk 6]
```diff
@@ -76,11 +76,13 @@ pub mod uploader;
 pub use logging::GitlabLayer;
 use tokio_util::sync::CancellationToken;
 use tracing::instrument::WithSubscriber;
+pub use uploader::UploadFile;
 mod runlist;
 use crate::runlist::RunList;
-use uploader::Uploader;
 
 use futures::prelude::*;
+use futures::AsyncRead;
```
Comment (sjoerdsimons) [inline gitlab-runner/src/lib.rs:84]: As we depend on Tokia should we use futres::AsyncRead or Tokio::AsyncRead ; did you pick one over the others for a specific reason ooi

File: gitlab-runner/src/lib.rs [hunk 7]
```diff
@@ -106,6 +108,52 @@ macro_rules! outputln {
...
+    ///
+    /// This is the path on disk from the root of the checkout for
+    /// Gitlab's own runner. It should match the paths that users are
+    /// expected to specify when requesting artifacts in their jobs.
+    fn get_path(&self) -> Cow<'_, str>;
+
+    /// Get something that can provide the data for this file.
+    ///
+    /// This can be any implementor of [`AsyncRead`].
+    fn get_data(&self) -> Self::Data<'_>;
```
[the remark this answers is not in the document]
Comment (eds-collabora) [reply gitlab-runner/src/lib.rs:135]: So the reason I didn't do that was to avoid the `async_trait` mess, since it's not necessary; the magical future methods (i.e. `flatten_stream`) make this relatively convenient and simple to implement, once you have the recipe. I.e. you only really need one level of "asyncness", you can hide any amount of additional layers of async just by flattening, which simplified the interface, and avoids making another `async_trait`

File: gitlab-runner/src/lib.rs [hunk 8]
```diff
@@ -132,14 +183,14 @@ pub trait CancellableJobHandler: Send {
...
+
+    /// Get a list of the files available to upload
     ///
-    /// This gets called depending on whether the job definition calls for artifacts to be uploaded
-    /// based on the result of the script run. If the job is cancelled while this is running, it
-    /// will still be run to completion. If the job was already cancelled, this will not be called.
-    async fn upload_artifacts(&mut self, _uploader: &mut Uploader) -> JobResult {
-        Ok(())
+    /// See the description [`UploadableFile`] for more information.
+    async fn get_uploadable_files(&mut self) -> Result<Box<dyn Iterator<Item = U> + Send>, ()> {
```
[the remark this answers is not in the document]
Comment (sjoerdsimons) [reply gitlab-runner/src/lib.rs:190]: Right i didn't think about not being able to have default for an associated type; You'd have to make it a parameter of the trait to mimic that which has it's own annoyances; 

So Yeah i'm fine with this for convenience purposes

File: gitlab-runner/src/lib.rs (hunk 6 above)
Comment (sjoerdsimons) [reply gitlab-runner/src/lib.rs:84]: Futures is meant to follow/proof out stuff that's going into stdlib as well indeed as ; But there were still quite some discussion around AsyncRead last i checked. 

I don't really have a strict preference here; there are also adapters both way around anyway; So i'm fine with going, use futures as external API where applicable (e.g. we use tokio for cancellation token as futures doesn't have that) :)

File: gitlab-runner/Cargo.toml [hunk 9]
```diff
@@ -12,6 +12,7 @@ readme = "../README.md"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
+flate2 = "1"
```
[the remark this answers is not in the document]
Comment (eds-collabora) [reply gitlab-runner/Cargo.toml:15]: I will tackle this in a follow-up.

File: gitlab-runner/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (sjoerdsimons) [reply gitlab-runner/src/lib.rs:135]: ah i didn't remember flatten_stream was a thing indeed :) 

I'm still a bit on the fence about this one because it means your Data type now has to be `FlattenStream<your data>`  and you'd probably do an async closure  or function to call flatten_stream on. Which seems a lot less pleasant for a user and given async_trait isn't introduced this for this avoiding the mess doesn't seem like a big benefit.  Oh also you'd have to be able to name the type so ; which you really can't do for functions or closures, which quickly forces things down the `Box<dyn> route`

Fwiw my hope is that in a year or two we'd be able to drop `async_trait` entirely and move to proper native async traits once Rust supports it; So i'm more treating `async_trait` as some annoying sprinklink you hav eto do for now to work around language limitations and ignoring that internally it does the same `Box<dyn>` tricks you're doing explicitly here :p

File: gitlab-runner-mock/src/job.rs [hunk 10]
```diff
@@ -211,9 +224,9 @@ impl MockJob {
         inner.log_patches
     }
 
-    pub fn artifact(&self) -> Arc<Vec<u8>> {
+    pub fn uploaded_artifacts(&self) -> impl Iterator<Item = MockUploadedJobArtifact> + '_ {
         let inner = self.inner.lock().unwrap();
-        inner.artifact.clone()
+        inner.uploaded_artifacts.clone().into_iter()
```
[the remark this answers is not in the document]
Comment (eds-collabora) [reply gitlab-runner-mock/src/job.rs:235]: I thought about that, but a convention of generally returning iterators is quite nice here, so I'll leave this as-is for now.
